File: ascon_api.c
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct{
    uint8_t *key;
    uint8_t *nonce;
    uint8_t *adata; 
    uint8_t  adlen;
    uint8_t *plaintext;
    uint8_t  ptlen;
    uint8_t *ciphertext;
    uint8_t *tag;
}ascon_aead_t;

// Utility Functions
uint64_t rotr(uint64_t x, int n) {
    return ((x >> n) | (x << (64 - n))) & 0xFFFFFFFFFFFFFFFFULL;
}

uint64_t bytes_to_int(const uint8_t* b) {
    uint64_t result = 0;
    for (int i = 0; i < 8; i++) {
        result = (result << 8) | b[i];
    }
    return result;
}

// ASCON Permutation Layers
uint64_t constant_layer(uint64_t x2, int rounds, int counter) {
    if (rounds == 6) {
        return x2 ^ (0x96 - (counter - 1) * 15);
    } else if (rounds == 8) {
        return x2 ^ (0xb4 - (counter - 1) * 15);
    } else {
        return x2 ^ (0xf0 - (counter - 1) * 15);
    }
}

void substitution_layer(uint64_t* state) {
    uint64_t x0 = state[0], x1 = state[1], x2 = state[2], x3 = state[3], x4 = state[4];
    x0 ^= x4; x4 ^= x3; x2 ^= x1;
    uint64_t t0 = (~x0) & x1;
    uint64_t t1 = (~x1) & x2;
    uint64_t t2 = (~x2) & x3;
    uint64_t t3 = (~x3) & x4;
    uint64_t t4 = (~x4) & x0;
    x0 ^= t1; x1 ^= t2; x2 ^= t3; x3 ^= t4; x4 ^= t0;
    x1 ^= x0; x0 ^= x4; x3 ^= x2; x2 = ~x2 & 0xFFFFFFFFFFFFFFFFULL;
    state[0] = x0; state[1] = x1; state[2] = x2; state[3] = x3; state[4] = x4;
}

void linear_layer(uint64_t* state) {
    state[0] ^= rotr(state[0], 19) ^ rotr(state[0], 28);
    state[1] ^= rotr(state[1], 61) ^ rotr(state[1], 39);
    state[2] ^= rotr(state[2], 1) ^ rotr(state[2], 6);
    state[3] ^= rotr(state[3], 10) ^ rotr(state[3], 17);
    state[4] ^= rotr(state[4], 7) ^ rotr(state[4], 41);
    for (int i = 0; i < 5; i++) {
        state[i] &= 0xFFFFFFFFFFFFFFFFULL;
    }
}

void ascon_permutation(uint64_t* state, int rounds) {
    for (int r = 1; r <= rounds; r++) {
        state[2] = constant_layer(state[2], rounds, r);
        substitution_layer(state);
        linear_layer(state);
        printf("state: %llx %llx %llx %llx %llx\n", state[0], state[1], state[2], state[3], state[4]);
    }
}

void ascon_encrypt(ascon_aead_t ascon_aead_t) {

    if (!ascon_aead_t.key || !ascon_aead_t.nonce || !ascon_aead_t.adata 
        || !ascon_aead_t.ciphertext || !ascon_aead_t.plaintext) return;

    // Initialization
    uint64_t state[5];
    state[0] = 0x80400c0600000000;
    state[1] = bytes_to_int(ascon_aead_t.key);
    state[2] = bytes_to_int(ascon_aead_t.key + 8);
    state[3] = bytes_to_int(ascon_aead_t.nonce);
    state[4] = bytes_to_int(ascon_aead_t.nonce + 8);
    ascon_permutation(state, 12);
    state[3] ^= bytes_to_int(ascon_aead_t.key);
    state[4] ^= bytes_to_int(ascon_aead_t.key + 8);

    uint8_t rate = 8;
    uint8_t s = ascon_aead_t.adlen / rate + 1; 
    uint8_t t = ascon_aead_t.ptlen / rate + 1;
    uint8_t l = ascon_aead_t.ptlen % rate;
    uint8_t *pad_ad = malloc(s * rate);
    uint8_t *pad_pt = malloc(t * rate);

    // padding associated data
    for (uint8_t i = 0; i < ascon_aead_t.adlen; i++)
        pad_ad[i] = ascon_aead_t.adata[i];
    pad_ad[ascon_aead_t.adlen] = 0x80;
    for (uint8_t i = ascon_aead_t.adlen + 1; i < (s * rate); i++)
        pad_ad[i] = 0x00;

    // padding plaintext
    for (uint8_t i = 0; i < ascon_aead_t.ptlen; i++)
        pad_pt[i] = ascon_aead_t.plaintext[i];
    pad_pt[ascon_aead_t.ptlen] = 0x80;
    for (uint8_t i = ascon_aead_t.ptlen + 1; i < (t * rate); i++)
        pad_pt[i] = 0x00;

    // absorb associated data
    for (uint8_t i = 0; i < s; i++) {
        state[0] ^= bytes_to_int(pad_ad + i * rate);  
        if (rate == 16)
            state[1] ^= bytes_to_int(pad_ad + i * rate + 8);
        ascon_permutation(state, 6);
    }
    state[4] ^= 1;
    free(pad_ad);

    // absorb ciphertext
    for (uint8_t i = 0; i < t-1; i++) {
        state[0] ^= bytes_to_int(pad_pt + i * rate);  
        if (rate == 16)
            state[1] ^= bytes_to_int(pad_pt + i * rate + 8);

        for (int j = 0; j < rate; j++) {
            ascon_aead_t.ciphertext[i * rate + j] = (state[0] >> (8 * (rate - 1 - j))) & 0xFF;
        }
        if (rate == 16) {
            for (int j = 0; j < rate; j++) {
                ascon_aead_t.ciphertext[i * rate + 8 + j] = (state[1] >> (8 * (rate - 1 - j))) & 0xFF;
            }
        }

        ascon_permutation(state, 6);
    }

    state[0] ^= bytes_to_int(pad_pt + (t - 1) * rate);
    if (rate == 16)
        state[1] ^= bytes_to_int(pad_pt + (t - 1) * rate + 8);

    for (int j = 0; j < l; j++) {
        ascon_aead_t.ciphertext[(t - 1) * rate + j] = (state[0] >> (8 * (rate - 1 - j))) & 0xFF;
    }
    free(pad_pt);

    // Finalization
    state[1] ^= bytes_to_int(ascon_aead_t.key);
    state[2] ^= bytes_to_int(ascon_aead_t.key + 8);
    ascon_permutation(state, 12);

    for (int i = 0; i < 8; i++) {
        ascon_aead_t.tag[i]     = ((state[3] >> (56 - 8 * i)) & 0xFF) ^ ascon_aead_t.key[i];
        ascon_aead_t.tag[i + 8] = ((state[4] >> (56 - 8 * i)) & 0xFF) ^ ascon_aead_t.key[i + 8];
    }
}

int main() {
    uint8_t key[16] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f};
    uint8_t nonce[16] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f};
    uint8_t ad[5] = {0x41, 0x53, 0x43, 0x4f, 0x4e};
    uint8_t pt[4] = {0x8b, 0x86, 0xd9, 0x32};
    uint8_t ct[4];
    uint8_t tag[16];

    ascon_aead_t ascon_aead = {
        .key        = key,
        .nonce      = nonce,
        .adata      = ad,
        .adlen      = sizeof(ad),
        .plaintext  = pt,
        .ptlen      = sizeof(pt),
        .ciphertext = ct,
        .tag        = tag
    };

    clock_t start = clock();
    ascon_encrypt(ascon_aead);
    clock_t end = clock();

    double elapsed = (double)(end - start) / CLOCKS_PER_SEC;
    printf("Time taken: %.6f seconds\n", elapsed);
    
    printf("Ciphertext: ");
    for (int i = 0; i < 4; i++) printf("%02X ", ct[i]);
    printf("\nTag: ");
    for (int i = 0; i < 16; i++) printf("%02X ", tag[i]);
    printf("\n");

    return 0;
}

File: ascon_api.py
# parameter
ASCON_AEAD_128  = 0
ASCON_AEAD_128A = 1

# --- Utility functions ---
def rotr(x: int, n: int) -> int:
    return ((x >> n) | ((x << (64 - n)) & 0xFFFFFFFFFFFFFFFF)) 

def bytes_to_int(b: bytes) -> int:
    return int.from_bytes(b[:8], byteorder='big')

# permutation layers
def constant_layer(x2, rounds, counter):
    if rounds == 6:
        return x2 ^ (0x96 - (counter - 1) * 15)
    elif rounds == 8:
        return x2 ^ (0xb4 - (counter - 1) * 15)
    else:
        return x2 ^ (0xf0 - (counter - 1) * 15)

def substitution_layer(state):
    x0, x1, x2, x3, x4 = state
    x0 ^= x4; x4 ^= x3; x2 ^= x1
    t0 = (~x0) & x1
    t1 = (~x1) & x2
    t2 = (~x2) & x3
    t3 = (~x3) & x4
    t4 = (~x4) & x0
    x0 ^= t1; x1 ^= t2; x2 ^= t3; x3 ^= t4; x4 ^= t0
    x1 ^= x0; x0 ^= x4; x3 ^= x2; x2 = ~x2 & 0xFFFFFFFFFFFFFFFF
    state[:] = [x0, x1, x2, x3, x4]

def linear_layer(state):
    state[0] ^= rotr(state[0], 19) ^ rotr(state[0], 28)
    state[1] ^= rotr(state[1], 61) ^ rotr(state[1], 39)
    state[2] ^= rotr(state[2], 1) ^ rotr(state[2], 6)
    state[3] ^= rotr(state[3], 10) ^ rotr(state[3], 17)
    state[4] ^= rotr(state[4], 7) ^ rotr(state[4], 41)
    for i in range(5):
        state[i] &= 0xFFFFFFFFFFFFFFFF

def ascon_permutation(state, rounds):
    for r in range(1, rounds + 1):
        state[2] = constant_layer(state[2], rounds, r)
        substitution_layer(state)
        linear_layer(state)

# --- AEAD encrypt
def  ascon_encrypt(variant: int, key: bytes, nonce: bytes, adata: bytes, adlen: int,
    plaintext: bytes, ptlen: int, ciphertext: bytearray, tag: bytearray):
    if variant == ASCON_AEAD_128:
        IV = 0x80400c0600000000
        round_a = 12
        round_b = 6
        rate = 8   # bytes (64-bit)
    elif variant == ASCON_AEAD_128A:
        IV = 0x80800c0800000000
        round_a = 12
        round_b = 8
        rate = 16  # bytes (128-bit)

    # Khởi tạo state
    state = [0]*5
    state[0] = IV
    state[1] = bytes_to_int(key[0:8])
    state[2] = bytes_to_int(key[8:16])
    state[3] = bytes_to_int(nonce[0:8])
    state[4] = bytes_to_int(nonce[8:16])

    ascon_permutation(state, round_a)

    state[3] ^= bytes_to_int(key[0:8])
    state[4] ^= bytes_to_int(key[8:16])

    # Tinh so khoi va padding
    s = (adlen + rate - 1) // rate
    t = (ptlen + rate - 1) // rate
    l = ptlen % rate

    pad_ad = bytearray(s * rate)
    pad_pt = bytearray(t * rate)

    # Padding associated data
    pad_ad[:adlen] = adata[:adlen]
    pad_ad[adlen] = 0x80

    # Padding plaintext
    pad_pt[:ptlen] = plaintext[:ptlen]
    pad_pt[ptlen] = 0x80

    # Absorb associated data
    for i in range(s):
        block = pad_ad[i*rate:(i+1)*rate]
        state[0] ^= bytes_to_int(block[0:8])
        if variant == ASCON_AEAD_128A:
            state[1] ^= bytes_to_int(block[8:16])
        ascon_permutation(state, round_b)

    state[4] ^= 1

    # Absorb plaintext
    for i in range(t - 1):
        block = pad_pt[i*rate:(i+1)*rate]
        state[0] ^= bytes_to_int(block[0:8])
        if variant == ASCON_AEAD_128A:
            state[1] ^= bytes_to_int(block[8:16])

        #  ciphertext 
        for j in range(rate):
            ciphertext[i*rate + j] = ((state[0] >> (8 * (rate - 1 - j))) & 0xFF)
            if variant == ASCON_AEAD_128A:
                ciphertext[i*rate + 8 + j] = ((state[1] >> (8 * (rate - 1 - j))) & 0xFF)

        ascon_permutation(state, round_b)

    # Xử lý khối cuối
    last_block = pad_pt[(t-1)*rate : t*rate]
    state[0] ^= bytes_to_int(last_block[0:8])
    if variant == ASCON_AEAD_128A:
        state[1] ^= bytes_to_int(last_block[8:16])

    for j in range(l):
        if j < rate:
            ciphertext[(t-1)*rate + j] = (state[0] >> (56 - 8*j)) & 0xFF
        else:
            ciphertext[(t-1)*rate + j] = (state[1] >> (56 - 8*(j - 8))) & 0xFF


    # Finalization
    if variant == ASCON_AEAD_128:
        state[1] ^= bytes_to_int(key[0:8])
        state[2] ^= bytes_to_int(key[8:16])
    else:  # ASCON_AEAD_128A
        state[2] ^= bytes_to_int(key[0:8])
        state[3] ^= bytes_to_int(key[8:16])

    ascon_permutation(state, round_a)

    # Tag 
    for i in range(8):
        tag[i]     = ((state[3] >> (56 - 8*i)) & 0xFF) ^ key[i]
        tag[i + 8] = ((state[4] >> (56 - 8*i)) & 0xFF) ^ key[i + 8]
        
    return ciphertext, tag

# --- AEAD decryption
# def ascon_decrypt(variant: int, key: bytes, nonce: bytes, adata: bytes, adlen: int,
#     ciphertext: bytes, ctlen: int, plaintext: bytearray, tag: bytearray):
#     if variant == ASCON_AEAD_128:
#         IV = 0x80400c0600000000
#         round_a = 12
#         round_b = 6
#         rate = 8   # bytes (64-bit)
#     elif variant == ASCON_AEAD_128A:
#         IV = 0x80800c0800000000
#         round_a = 12
#         round_b = 8
#         rate = 16  # bytes (128-bit)

#     # Khởi tạo state
#     state = [0]*5
#     state[0] = IV
#     state[1] = bytes_to_int(key[0:8])
#     state[2] = bytes_to_int(key[8:16])
#     state[3] = bytes_to_int(nonce[0:8])
#     state[4] = bytes_to_int(nonce[8:16])

#     ascon_permutation(state, round_a)

#     state[3] ^= bytes_to_int(key[0:8])
#     state[4] ^= bytes_to_int(key[8:16])

#     # Tinh so khoi va padding
#     s = (adlen + rate - 1) // rate
#     t = (ctlen + rate - 1) // rate
#     l = ctlen % rate

#     pad_ad = bytearray(s * rate)
#     pad_ct = bytearray(t * rate)

#     # Padding associated data
#     pad_ad[:adlen] = adata[:adlen]
#     pad_ad[adlen] = 0x80

#     # Padding plaintext
#     pad_ct[:ctlen] = ciphertext[:ctlen]
#     pad_ct[ctlen] = 0x80

#     # Absorb associated data
#     for i in range(s):
#         block = pad_ad[i*rate:(i+1)*rate]
#         state[0] ^= bytes_to_int(block[0:8])
#         if variant == ASCON_AEAD_128A:
#             state[1] ^= bytes_to_int(block[8:16])
#         ascon_permutation(state, round_b)

#     state[4] ^= 1

#     # Absorb plaintext
#     for i in range(t - 1):
#         block = pad_ct[i*rate:(i+1)*rate]
#         C0 = bytes_to_int(block[0:8])
#         P0 = state[0] ^ C0
#         for j in range(8):
#             plaintext[i*rate + j] = (P0 >> (56 - 8*j)) & 0xFF

#         if variant == ASCON_AEAD_128A:
#             C1 = bytes_to_int(block[8:16])
#             P1 = state[1] ^ C1
#             for j in range(8):
#                 plaintext[i*rate + 8 + j] = (P1 >> (56 - 8*j)) & 0xFF

#         state[0] = C0
#         if variant == ASCON_AEAD_128A:
#             state[1] = C1

#         ascon_permutation(state, round_b)

#     # Xử lý khối cuối
#     last_block = pad_ct[(t-1)*rate : t*rate]
#     C0 = bytes_to_int(last_block[0:8])
#     P0 = state[0] ^ C0
#     if variant == ASCON_AEAD_128A:
#         C1 = bytes_to_int(last_block[8:16])
#         P1 = state[1] ^ C1

#     for j in range(l):
#         if j < rate:
#             plaintext[(t-1)*rate + j] = (P0 >> (56 - 8*j)) & 0xFF
#         else:
#             plaintext[(t-1)*rate + j] = (P1 >> (56 - 8*(j - 8))) & 0xFF

#     state[0] = C0
#     if variant == ASCON_AEAD_128A:
#         state[1] = C1  

#     # Finalization
#     if variant == ASCON_AEAD_128:
#         state[1] ^= bytes_to_int(key[0:8])
#         state[2] ^= bytes_to_int(key[8:16])
#     else:  # ASCON_AEAD_128A
#         state[2] ^= bytes_to_int(key[0:8])
#         state[3] ^= bytes_to_int(key[8:16])

#     ascon_permutation(state, round_a)

#     # Tag 
#     for i in range(8):
#         tag[i]     = ((state[3] >> (56 - 8*i)) & 0xFF) ^ key[i]
#         tag[i + 8] = ((state[4] >> (56 - 8*i)) & 0xFF) ^ key[i + 8]
        
#     return plaintext, tag
def ascon_decrypt(variant: int, key: bytes, nonce: bytes, adata: bytes, adlen: int,
                  ciphertext: bytes, ctlen: int, plaintext: bytearray, tag: bytearray):
    if variant == ASCON_AEAD_128:
        IV = 0x80400c0600000000
        round_a = 12
        round_b = 6
        rate = 8   # bytes (64-bit)
    elif variant == ASCON_AEAD_128A:
        IV = 0x80800c0800000000
        round_a = 12
        round_b = 8
        rate = 16  # bytes (128-bit)

    # Initialize state
    state = [0] * 5
    state[0] = IV
    state[1] = bytes_to_int(key[0:8])
    state[2] = bytes_to_int(key[8:16])
    state[3] = bytes_to_int(nonce[0:8])
    state[4] = bytes_to_int(nonce[8:16])

    ascon_permutation(state, round_a)

    state[3] ^= bytes_to_int(key[0:8])
    state[4] ^= bytes_to_int(key[8:16])

    # Calculate number of blocks
    s = (adlen + rate - 1) // rate
    t = (ctlen + rate - 1) // rate
    l = ctlen % rate

    # Padding associated data
    pad_ad = bytearray(s * rate)
    pad_ad[:adlen] = adata[:adlen]
    pad_ad[adlen] = 0x80

    # Absorb associated data
    for i in range(s):
        block = pad_ad[i * rate:(i + 1) * rate]
        state[0] ^= bytes_to_int(block[0:8])
        if variant == ASCON_AEAD_128A:
            state[1] ^= bytes_to_int(block[8:16])
        ascon_permutation(state, round_b)

    state[4] ^= 1

    # Absorb ciphertext 
    for i in range(t - 1):
        block = ciphertext[i * rate:(i + 1) * rate]
        C0 = bytes_to_int(block[0:8])
        P0 = state[0] ^ C0
        for j in range(8):
            plaintext[i * rate + j] = (P0 >> (56 - 8 * j)) & 0xFF

        if variant == ASCON_AEAD_128A:
            C1 = bytes_to_int(block[8:16])
            P1 = state[1] ^ C1
            for j in range(8):
                plaintext[i * rate + 8 + j] = (P1 >> (56 - 8 * j)) & 0xFF

        state[0] = C0
        if variant == ASCON_AEAD_128A:
            state[1] = C1

        ascon_permutation(state, round_b)

    # Process last block
    last_block = ciphertext[(t - 1) * rate: t * rate]
    C0 = bytes_to_int(last_block[0:8])
    P0 = state[0] ^ C0
    if variant == ASCON_AEAD_128A:
        C1 = bytes_to_int(last_block[8:16])
        P1 = state[1] ^ C1

    for j in range(l):
        if j < 8:
            plaintext[(t - 1) * rate + j] = (P0 >> (56 - 8 * j)) & 0xFF
        elif variant == ASCON_AEAD_128A:
            plaintext[(t - 1) * rate + j] = (P1 >> (56 - 8 * (j - 8))) & 0xFF

    pad_last = bytearray(rate)
    pad_last[:l] = plaintext[(t - 1) * rate:(t - 1) * rate + l]
    pad_last[l] = 0x80
    state[0] ^= bytes_to_int(pad_last[0:8])
    if variant == ASCON_AEAD_128A:
        state[1] ^= bytes_to_int(pad_last[8:16])

    # Finalization
    if variant == ASCON_AEAD_128:
        state[1] ^= bytes_to_int(key[0:8])
        state[2] ^= bytes_to_int(key[8:16])
    else:  # ASCON_AEAD_128A
        state[2] ^= bytes_to_int(key[0:8])
        state[3] ^= bytes_to_int(key[8:16])

    ascon_permutation(state, round_a)

    # Generate tag
    for i in range(8):
        tag[i] = ((state[3] >> (56 - 8 * i)) & 0xFF) ^ key[i]
        tag[i + 8] = ((state[4] >> (56 - 8 * i)) & 0xFF) ^ key[i + 8]

    return plaintext, tag
def main():
    key   = bytes([0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0a,0x0b,0x0c,0x0d,0x0e,0x0f])
    nonce = bytes([0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0a,0x0b,0x0c,0x0d,0x0e,0x0f])
    ad    = bytes([0x41,0x53,0x43,0x4f,0x4e])
    pt    = bytes([0x8b,0x86,0xd9,0x32,0x8b,0x86,0xd9,0x32])
    ct    = bytearray(len(pt))
    tag   = bytearray(16)

    ascon_encrypt(ASCON_AEAD_128A, key, nonce, ad, len(ad), pt, len(pt), ct, tag)

    print("Ciphertext:", " ".join(f"{b:02X}" for b in ct))
    print("Tag:", " ".join(f"{b:02X}" for b in tag))

    pt_dec  = bytearray(len(ct))
    tag_dec = bytearray(16)
  
    ascon_decrypt(ASCON_AEAD_128A, key, nonce, ad, len(ad), ct, len(ct), pt_dec, tag_dec)
    print("Plaintext:", " ".join(f"{b:02X}" for b in pt_dec))
    print("Tag:", " ".join(f"{b:02X}" for b in tag_dec))

if __name__ == "__main__":
    main()
